### ISSUE TITLE
Realm: Compact Sparse Instances

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#624]
Realm should support the ability to create "compressed" instances for index spaces with sparsity maps so that we can make instances that take up the minimal amount of memory possible with the understanding that the accessors for these instances will likely not support O(1) memory accesses, but likely something that is O(log N)-ish on the number of rectangles in the sparsity map, possibly using a K-D tree like data structure. Needs to work on GPUs too. DMA system needs to know how to copy to/from them efficiently as well.

**Comment by streichler** [comment]
I doubt we really want an accessor that's more than amortized O(1).  I think we instead want a way to iterate coarsely over affine pieces of an instance and still use affine accessors within a piece.  I'd like to sit down and brainstorm with @lightsighter , @elliottslaughter , @magnatelee , and anybody else who's interested to come up with a plan that includes how C++ and Regent applications will use this.

**Comment by mariodirenzo** [comment]
I'm encountering this issue in my solver. If I have a periodic domain, the partitions that I am defining to support the stencil operations wrap around the main region. The result is that the nodes with the periodic boundaries allocate a lot of memory. It would be good to able to obtain instances with the same size as the partitions that I am defining in the solver

**Comment by streichler** [comment]
@mariodirenzo what would you like the inner loop of your task to look like?  how do you want to compute addresses that wrap around in the index space?

**Comment by mariodirenzo** [comment]
If I understand your question correctly, the inner loop of the task looks like
```
for c in r do
    r[c].b = r[c].a + r[r[c].cp1].a
end
```
where `r[c].cp1` is an int3d stored in each point that on the boundaries of the domain is 
- `r[c].cp1 = c` if it is not periodic
- `r[c].cp1 = (c + int3d(1,0,0))%r.bounds` if it is periodic.

**Comment by streichler** [comment]
Right, so the for loop will turn into something like:
```
for(c.x = ...; c.x < ...; c.x++)
  for(c.y = ...; c.y < ...; c.y++)
    for(c.z = ...; c.z < ...; c.z++) {
```
and an affine access like `r[c].b` will turn into:
```
*(b_base + c.x * A + c.y * B + c.z * C)
```
Nice and efficient.  However, it gets ugly as soon as there are multiple pieces in an instance.
If we specialize a lot in the C++ (i.e. build an accessor designed for this wraparound case and little else), we can probably get to something like:
```
*((c.x < G) ? (b_base_1 + c.x * A + ...) : (b_base_2 + c.x * D + ...)
```
and then if we're lucky, the compiler might be able to pull that conditional through two of the loops.

The alternative is for the `for c in r` construct to be smarter and first split `r` into subrectangles where for which every access in that subrectangle is affine.  @magnatelee can you comment on whether the parallelizer analysis might be able to help with this?

**Comment by mariodirenzo** [comment]
Considering that the partitions that I create are not completely sparse but they just select a number of contiguous elements from the other side of the domain, can't you just create an affine piece that accommodates on one side the copies that come from the other side of the domain.
To give a practical example, if I have a 1d region of size 256 and a partition that contains the indices [191:255] and [0]. Could you map this partition on an instance of size 66, where the first 65 elements of the instance contain  [191:255] and the last contains [0]?
Probably this is much less general than what you have described but, it would work in my case.

**Comment by streichler** [comment]
> To give a practical example, if I have a 1d region of size 256 and a partition that contains the indices [191:255] and [0]. Could you map this partition on an instance of size 66, where the first 65 elements of the instance contain [191:255] and the last contains [0]?

Yes, that's basically how the instance will be laid out in memory.  The question is how we convert
```r.a[x]``` -> ```*((inst_base - 191) + x)```
for `x >= 191` and
```r.a[x]``` -> ```*(inst_base + 65) + x)```
for `x == 0`.

**Comment by magnatelee** [comment]
@streichler The parallelizer can certainly generate loops to eliminate the inclusion checks against different stencil regions, but we're not doing it at the moment because of potential combinatorial explosion of cases; i.e., in general we need to generate 2^N loops for N stencils in a parallelizable loop. For fully affine cases, the parallelizer instead generates a task variant requiring the interior and stencil regions to be collocated in an instance in order to enjoy contiguous access. Mario's case above won't benefit from this optimization as his stencil cannot be proven affine at compile time.

**Comment by streichler** [comment]
@mariodirenzo do you need the full generality of the `cp1` field in your example, or would it have been possible (either manually or via metaprogramming) to have two versions that explicitly do either:
```r[c].b = r[c].a + r[(max(c.x + 1, 255), c.y, c.z)].a``` (for nonperiodic)
or
```r[c].b = r[c].a + r[(c + int3d(1,0,0))%r.bounds].a``` (for periodic)
?

@magnatelee if the regent code did look like that, would the stencil analysis be able to generate accessors that bake the structured indirection into the accessor using an interface similar to this:
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/inst_layout.h#L314-320
(a more general interface that allows for "sort of affine" transforms like the periodic wrapping case is in the works)

Also @magnatelee but @elliottslaughter too: Is there any control right now for what kind of accessors are created in a regent task or is it hardcoded for affine right now?  If the latter, can we figure out a syntax that would let application code suggest an alternative (e.g. the GenericAccessor for arbitrary wraparound cases or perhaps a newer TwoAffinePiecesAccessor that had the functionality I talked about above)?

**Comment by magnatelee** [comment]
@streichler I believe Mario wouldn't need to clip the coordinate as he is adding some pads for boundary conditions and iterating over only the interior elements.

The parallelizer should be able to generate such accessors as it already has accurate information about the shape of stencils. I guess the accessors for periodic wrapping cases would need to know the layout of instances to figure out the wrapped parts, but it's not 100% clear to me.

All accessors created by the compiler are hardcoded for affine right now. Adding a syntax to specify the accessor type sounds fine to me, and I guess we can even automate for some cases based on the type of indices, but I'm not sure how we can make those accessors play nicely with GPU code generation. Will Realm give us back pointers to device-side accessors?

**Comment by streichler** [comment]
> but I'm not sure how we can make those accessors play nicely with GPU code generation. Will Realm give us back pointers to device-side accessors?

The intent is that accessors with GPU support are trivially copyable so that they can be constructed on the host and then passed to a CUDA kernel by value.  `AffineAccessor` has this property, but `GenericAccessor` does not (and therefore isn't supported on the GPU).

**Comment by magnatelee** [comment]
>The intent is that accessors with GPU support are trivially copyable so that they can be constructed on the host and then passed to a CUDA kernel by value. AffineAccessor has this property, but GenericAccessor does not (and therefore isn't supported on the GPU).

Affine accesses are what the compiler can handle in GPU code generation as well. I was hoping that at least some kinds of accessors other than AffineAccessor that can handle "chunky" sparse instances are trivially copyable, or at least made copyable with some effort. It would be even nicer if that chunkiness could be decided by the mapper so that the mapper can guide which kind of accessors the compiler needs to emit in the task variant. We are kind of doing this implicitly; we always create instances having a single chunk, which allows the compiler to always emit affine accessors. An obvious generalization of this would be to take a number of chunks. There is still a question of how to construct those chunks in cases where there are more dense pieces than requested chunks.

**Comment by elliottslaughter** [comment]
Just to be clear though, we're not using the C++ types at all (and I don't think we can). We've reimplemented the affine accessor in Regent; we'd need to also implement any other kinds of accessors as well.

One issue with "chunky" accessors is uncentered accesses (i.e., `a[b[x]]` for some arbitrary field `b` or `a[f(x)]` for some affine function `f`). I don't remember ever getting to a satisfying resolution about this in any of our designs.

**Comment by mariodirenzo** [comment]
> would it have been possible (either manually or via metaprogramming) to have two versions that explicitly do either
> r[c].b = r[c].a + r[(max(c.x + 1, 255), c.y, c.z)].a (for nonperiodic)
> or
> r[c].b = r[c].a + r[(c + int3d(1,0,0))%r.bounds].a (for periodic)
> ?

The current version of the code works in a way that the access to the stencil points is modified though the field `cp1` (actually goes up to `cp3` for the +3 access) in case of non-periodic boundary conditions in a way such that something equivalent to `max(c.x + 1, 255)` is enforced, ghost region do not wrap around the region and I do not see this issue.
I could metaprogram a specific version for periodic access of the stencil if it is necessary.

**Comment by streichler** [comment]
In practice, what do you think a reasonable bound on the number of chunks you'd need is?

> There is still a question of how to construct those chunks in cases where there are more dense pieces than requested chunks.

Well, it's an optimization problem with an easily-specified cost metric.  I agree the heuristics/solution for it might be complicated though.

> One issue with "chunky" accessors is uncentered accesses (i.e., a[b[x]] for some arbitrary field b or a[f(x)] for some affine function f). I don't remember ever getting to a satisfying resolution about this in any of our designs.

The pointer math I gave above works fine for uncentered accesses too.  There is an additional optimization opportunity for the structured indirect accesses though, and I'd like to make sure we take advantage of it as part of this rework.

**Comment by magnatelee** [comment]
> In practice, what do you think a reasonable bound on the number of chunks you'd need is?
I'm hoping that it won't exceed (cache line size)/8, otherwise the array that we store base pointers in won't be fetched as nicely.

> Well, it's an optimization problem with an easily-specified cost metric. I agree the heuristics/solution for it might be complicated though.
Yup, the problem can be easily stated, but the solution may not. I guess equal distribution of pieces across chunks without any optimization would get us something working initially. In the worst case where one of the bounding boxes blows up, we can still increase the number of chunks and hope for the best.

**Comment by mariodirenzo** [comment]
Is there any workaround that I can implement in the solver on in its mapper to momentarily avoid this issue until it is fixed?

**Comment by streichler** [comment]
The two options available to you right now (which are basically what other, more explicit, implementations do):
1. Use multiple region requirements, one for the contiguous "center" data, and one for each ghost region that might be wrapped around to the other side of the parent region.  This lifts the "which piece does my uncentered access point into" decision all the way into the application code, causing nasty-looking and slower code.
2. Bloat your region by the size of the largest stencil and explicitly copy the data on each edge to the opposite side so that you can leave the `%` out of the coordinate calculation.  This makes the stencil code as fast as possible (back to a single affine piece with no branching in the address calculation) - possibly faster than the best thing we'd ever be able to do with multi-piece instances.  It does add some complication to the top-level task, which must now be aware of the stencil pattern(s) of each of the tasks it calls.

I'd suggest option (2), and if it works well, we might be able to get the parallelizer to learn how to perform that trick automatically.

**Comment by elliottslaughter** [comment]
For what it's worth, (2) is what we do in the stencil code today, and it seems to work reasonably well. Though it does break the abstraction of having only a single global index space. (I mean, you still have a single index space, but now it's bloated with all these extra cells that don't belong to the original domain.)

**Comment by mariodirenzo** [comment]
Thanks for your reply!
Yes, (2) is probably the cleanest way of dealing with this issue.
It is also what I would do in a standard MPI code in order to enforce periodicity (I used to treat those points exactly as interprocessor points with a send and receive).
I'm trying to find a way to do it without rewriting the rest of the code.
Is it ok if I define the current root region as a subregion of the bloated region?

**Comment by lightsighter** [comment]
In addition to the "unstructured" case where we break down the accessor into dealing with "chunk" rectangles, I think there's another "strucutured" case that we should have here which is the case where the user can assert that the points in the index space are defined by an affine mapping and the range of that mapping is a dense rectangle. If the user provides this mapping and the range rectangle at construction time, then Realm should be able to create a dense allocation of memory which still has O(1) memory access time.

**Comment by streichler** [comment]
@lightsighter can you give a concrete example of this?  I think this case may already be supported.

**Comment by lightsighter** [comment]
```
a = np.array(range(0,8), np.int)
evens = a[0:7:2]
odds = a[1:8:2]
```

The instance for `evens` should by 32 bytes and the instance for `odds` should be 32 bytes.

**Comment by streichler** [comment]
What is the index space for `evens`?  `{ 0, 2, 4, 6 }` or `{ 0, 1, 2, 3 }`?

**Comment by lightsighter** [comment]
`{0, 2, 4, 6}` It's created by a partition-by-restriction call.

**Comment by bandokihiro** [comment]
Since @streichler asked me to scream loud, I would benefit a lot from this feature. My use case is the general unstructured one. Let me know if you need further details.

**Comment by lightsighter** [comment]
Work on this issue from the Legion side is being tracked by this [merge request](https://github.com/StanfordLegion/legion/issues/624)

**Comment by lightsighter** [comment]
A question for everyone here that has requested this feature: let's assume you're making a compact instance with "SOA" layout for two fields `A` and `B`. Assuming that our sparse index space for this instance has two rectangles of points `r1` and `r2`, which of the following layouts does that mean for you? (I'm listing things in the order they would exist in memory)

Layout 1: elements of r1:A, elements of r1:B, elements of r2:A, elements of r2:B
Layout 2: elements of r1:A, elements of r2:A, elements of r1:B, elements of r2:B

Can you see uses for both or do you always expect one of these?

**Comment by streichler** [comment]
I vote for "SOA" meaning "all of field A then all of field B".

**Comment by magnatelee** [comment]
I'll always want layout 2, as layout 1 isn't really "compact" for either of the fields. Unless the user specified otherwise, a compact instance with SOA layout should mean all values of each field are contiguously located in the instance.

**Comment by lightsighter** [comment]
Just to be clear, the use case for layout 1 is when you write code the looks like:

`
for rect in domain:
        for point in rect:
              access fields A[point] and B[point]
`

Seems like layout 1 would give you better cache locality in that case, whereas layout 2 would give you better cache locality for code that looks like

`
for point in domain:
     access A[point] and B[point]
`

**Comment by bandokihiro** [comment]
I think I see what @lightsighter wants to say. It becomes ambiguous since you have 3 levels, the recangle, the point and the field. Layout 1 can be interpreted as SOA if you define the array to be "everything related to one rectangle". But if talking about the SOA layout of a field, then I would agree that layout 2 is what I would think of an SOA layout.

**Comment by magnatelee** [comment]
@lightsighter I think your first example enjoys better cache locality only when sub-rectangles in the sparse region are very small, and if that's the kind of locality the code really wants to exploit, it should use AOS layout instead.

**Comment by lightsighter** [comment]
Ok, and now for the fun part: what happens when the dimension ordering is `DIM_X, DIM_F, DIM_Y`?

**Comment by lightsighter** [comment]
Here's another way to think about the problem: assume our field constraints had a `DIM_P` which is the "piece" dimension, that can be used to specify how the pieces should be laid out relative to the other dimensions like the spatial dimensions and the field dimensions. So far we have said:

AOS: DIM_F, DIM_X, DIM_Y, DIM_P
SOA: DIM_X, DIM_Y, DIM_P, DIM_F

If the user specifies `DIM_X, DIM_F, DIM_Y` then going off the SOA pattern would be `DIM_X, DIM_P, DIM_F, DIM_Y`, but going off the AOS pattern would be `DIM_X, DIM_F, DIM_Y, DIM_P`. You could also imagine wanting `DIM_X, DIM_F, DIM_P, DIM_Y`. I'm partial to `DIM_X, DIM_P, DIM_F, DIM_Y` since it would make my life easier, but that's hardly ever the right answer for how something should be implemented.

**Comment by alexaiken** [comment]
At the risk of opening a can of worms, I'm less interested in which one we call "SOA" in the blocked/pieced case than in having the ability to choose among the full set of layouts implied by the different possible orders of fields, pieces, and spatial dimensions.  While many of these aren't obviously useful on their own, in applications with long pipelines of tasks  layout conversions between tasks in the chain will likely be needed for the best performance, and some of the less obvious choices might turn out to be the best in specific circumstances.  But it sounds like we aren't considering supporting all layouts at this time?

Alex

________________________________

**Comment by lightsighter** [comment]
So if we were to go that direction, then we would need to add a new kind of dimension (call it `DIM_P` like my previous post) so that users can say how to order the pieces with respect to the other spatial dimensions and the field dimension. If we go this direction, I would require that all users must explicitly specify full dimension ordering for compact instances and that this be an entry in that list. If we go down this road though we should also consider whether or not to expose the ordering of the actual rectangles in the layout to the user as well.

**Comment by alexaiken** [comment]
I assume that if we went in this direction we would also give names to a handful of the most important/common orderings (SOA, AOS --whatever we decide they mean!), so that users didn't need to learn about ordering of dimensions unless they wanted to.

Alex

________________________________

**Comment by lightsighter** [comment]
Someone could build that library on top of the existing layout constraint interface. I haven't bothered to do that since most people just sit on top of the default mapper which just guesses at the layout. The only people that seem to override the default mapper to pick a layout want the lower level interface anyway.

**Comment by streichler** [comment]
A question for potential users of this feature.  We're looking at adding some controls that the mapper can use to control the coarsening of an instance (i.e. using fewer pieces at the cost of some storage overhead in wasted space).  Current bloated instances represent one extreme (i.e. fewest possible pieces) while "precise" sparse instances would represent the other, dedicating no storage to indices that are not in the index space.  To allow choices in the middle, the plan is to have two knobs:
(1) max piece count for the instance, and
(2) max storage overhead
If both constraints are used, it's possible that no instance can be constructed at all.

My question(s) are:
(a) does this seem reasonable, and
(b) what units should the storage overhead be measured in?  number of wasted elements?  overhead ratio as a percentage (i.e. 0=no overhead, 100 = no more than twice the necessary storage can be used)

**Comment by mariodirenzo** [comment]
From my point of view,
> (a) does this seem reasonable

yes, as long as specifying the constraint (1) is optional. That constraint is very easy to set in a structured case but might be impossible for unstructured.

> (b) what units should the storage overhead be measured in? number of wasted elements? overhead ratio as a percentage (i.e. 0=no overhead, 100 = no more than twice the necessary storage can be used)

I vote for the overhead ratio as a percentage. A threshold on that values looks much more general to me than the number of wasted elements.

**Comment by streichler** [comment]
> yes, as long as specifying the constraint (1) is optional. That constraint is very easy to set in a structured case but might be impossible for unstructured.

Both constraints are optional, but don't think of (1) as only being useful for structured.  It is meaningful in all cases (i.e. we currently use a constraint of 1 piece for both structured and unstructured) and it impacts the cost of the search to figure out which piece a given index lies in.  (Assuming some sort of acceleration structure, the cost includes a term that is logarithmic in the number of pieces.)

**Comment by mariodirenzo** [comment]
>  but don't think of (1) as only being useful for structured

I agree with you, I did not mean that it is only useful in structured. I was trying to say that setting a hard constraint on the number of pieces in an unstructured graph, where we do not know the topology, might be problematic and require a lot of tuning. But if the constraint is optional, then the observation does not apply.

**Comment by bandokihiro** [comment]
I agree with what have been said.

**Comment by rupanshusoi** [comment]
For the record, my use case will significantly benefit from this change. I have sparse partitions of ghost points that are constructed by image/preimage and set operations in Regent. CUDA tasks using these regions quickly run out of memory on increasing the number of sub-regions, limiting the size of input grids that can be run.

**Comment by lightsighter** [comment]
All the Legion APIs for compact instances have implementations in the `newinsts` branch, however many of them are still unusable because we're waiting on implementations of a few Realm features. 

First, the Legion `FieldAccessor` class can now be specialized with `Realm::MultiAffineAccessor` which supports instances that have multiple pieces. You can use that if you want to just access the instance directly like you do today, but it will not have O(1) cost (@streichler will need to say what the complexity will be as a function of the number of pieces).

Next, if you want to iterate over the pieces in your instance, you can use the new `PieceIterator` or `PieceIteratorT` classes to enumerate the rectangle bounds of each of the pieces. You can either iterate over the physical pieces in the instance, or over just the logical pieces on which the `PhysicalRegion` object has privileges.
https://gitlab.com/StanfordLegion/legion/-/blob/newinsts/runtime/legion.h#L2565-2640

Alternatively, if all you want to do is walk over the dense "spans" of a physical region, potentially spanning multiple pieces then you can use a `SpanIterator` class. This will return spans of elements with a uniform stride in the instance, potentially crossing multiple pieces if possible, so that you can iterate them quickly in memory. The `SpanIterator` class is here:
https://gitlab.com/StanfordLegion/legion/-/blob/newinsts/runtime/legion.h#L2642-2686
and the `Span` objects that it returns can be found here:
https://gitlab.com/StanfordLegion/legion/-/blob/newinsts/runtime/legion/legion_domain.h#L620-743

Lastly, you can now create instances in the mapper using the `LEGION_COMPACT_SPECIALIZE` value or the `LEGION_COMPACT_REDUCTION_SPECIALIZE` values which will signal to the runtime to create a compact instance.
https://gitlab.com/StanfordLegion/legion/-/blob/newinsts/runtime/legion/legion_config.h#L1910-1912
You can control the number of rectangles in these compacted instances with the `SpecializedConstraint` objects that now support specifying both the maximum number of rectangles and the maximum amount of overhead permitted when creating the instances:
https://gitlab.com/StanfordLegion/legion/-/blob/newinsts/runtime/legion/legion_constraint.h#L366-367
https://gitlab.com/StanfordLegion/legion/-/blob/newinsts/runtime/legion/legion_constraint.h#L393-394

Please review these APIs. If you want to propose changes or additional extensions now is the time to speak up.

**Comment by rupanshusoi** [comment]
@lightsighter How would one go about using this from Regent? I don't need to iterate, just access elements (via their index) in a sparse region of ghost points.

**Comment by lightsighter** [comment]
I'll let @elliottslaughter comment on plans for Regent support.

**Comment by elliottslaughter** [comment]
We'll need to think some more about Regent support. Regent currently makes assumptions (like having one base pointer for the whole region) when it's creating accessors. We'll need to add support in the compiler for these more sophisticated accessors. It's probably worth a separate issue since I think there will be issues to solve beyond those faced by Legion/Realm.

**Comment by rupanshusoi** [comment]
@elliottslaughter I'd be happy to help with this, partly because it directly affects my use case.

**Comment by elliottslaughter** [comment]
Ok, let's start a separate issue to talk about the Regent design and implementation.

**Comment by lightsighter** [comment]
Regent support being tracked by issue #878

**Comment by lightsighter** [comment]
We have our first working example of compact instances in Legion. I've modified the C++ version of Pennant to use compact instances to represent the aggregation of the "private", "shared", and "ghost" regions for each piece of the mesh. The code is in the `compact` branch at this repo (unfortunately this repo is private so if you want access request it from @gshipman):
https://github.com/gshipman/pennant-legion/tree/compact/

Note that in single node cases, you need to disable this optimization in the Pennant mapper that just makes a giant instance for all the points:
https://github.com/gshipman/pennant-legion/blob/compact/src/PennantMapper.cc#L547-L561

We make compact instances in the Pennant mapper here (you can play around with the "clumping" effect by changing `overhead` to be greater than zero.
https://github.com/gshipman/pennant-legion/blob/compact/src/PennantMapper.cc#L612-L614

There are lots of places now that use multi-region multi-affine accessors for indexing into compact instances for several logical regions at a time. You can grep for them with `AccessorMC`:
https://github.com/gshipman/pennant-legion/blob/compact/src/MyLegion.hh#L251-L254

**Comment by streichler** [comment]
Can we do some runs to get a sense of what the memory footprint vs execution time tradeoff is for pennant?

**Comment by lightsighter** [comment]
I have other things that I need to be doing. I can tell someone else how to do that though if they want to gather the data. I actually think the more important experiment is just demonstrating that weak scaling does not run out of memory anymore because of exceedingly large and sparse ghost instances as the number of nodes increases.

**Comment by lightsighter** [comment]
Support for these features in Legion have now landed in the `master` branch and have been upstreamed into the `control_replication` branch so I am removing my assignment.
